---
output: word_document
---
<!--
%\VignetteEngine{knitr::rmarkdown}
%\VignetteIndexEntry{How-to-use-the-EUFLOW-package}
-->


---
title: "EUFLOW: Expected-Utility-based Workflow Evaluation"
author: "Kevin K. McDade"
date: "`r Sys.Date()`"
vignette: >
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteKeyword{EUFLOW}
  %\SweaveUTF8
  %\usepackage[utf8](inputenc)
output: 
    rmarkdown::html_vignette:
        toc: yes
---

```{r setup, include=FALSE}
library(knitr)
knitr::opts_chunk$set(echo = TRUE)
library("EUFLOW")
options(stringsAsFactors = FALSE)
system("touch I_ran_Rmd")
```
### I. Purpose of the EUFLOW package

The data in bioinformatics is often in some “raw” form which is not yet ready for analysis. Processing this data often involves several steps, called variously a workflow, pipeline, or protocol. We have developed the EUFLOW package to assist users who have multiple versions of data that arise from different *workflow paths*. Each of the *final processed data* files should contain similar features as the rows of the dataframe with columns of identical samples. The collective *workflow path data* is refered to as the *Evaluation Dataset*.  In order to evaluate the alternitive versions of the data the user must also provide a *Reference Dataset*. The Evaluation Dataset and the Reference Dataset must have an expected quantitative relationship, such as mRNA/protein coorelation. The EUFLOW package utilizes this relationship (*Model Quality*) through our methodology to determine the *Expected Utility* for each of the *workflow paths*.  

### II. Usage of EUFLOW

Users are required to input two separate data files, regardless of the analysis type. The EUFLOW package depends upon a model quality heuristic which we will demonstate here as correlation between gene expression and protein expression. The first data file, for example, we will use ovarian TCGA RNASeq data on two separate workflows provided on the same samples. Only samples for which protein expression data was available are used in the RNASEQDATA file, the HUGO Gene names and descriptions for these proteins are presented in the table below. 

```{r}
data(RPPADATA.original)
RPPADATA<-RPPADATA.original
RPPADATA$X
```

A small slice of this data file (RNASEQDATA) demonstrates the data structure of numbered row names followed by a column of identifiers including a tag _v1 and _v2. The same samples were processed with two different workflow options RnaSeqv1 and RnaSeqv2. The output below shows the gene expression values for the first 9 identifiers (in this case gene names) with a tag to distinguish workflow option. For brevity only 4 of the 198 sample identifiers are represented.


```{r}
data(RNASEQDATA)
RNASEQDATA[1:9,1:5]
data(RNASEQV1v2PICCOLO)
#OVPICCOLO<-log2(OVCancerFeatureCountsDATA)
#OVPICCOLO[1:9,1:5]
RNASEQV1v2PICCOLO[1:9,1:5]

```



In the second data file we will use TCGA protein expression data on the samples. As above we will output only a subset. Alos note that as this is the reference dataset in this example only one identifier is represented rather than multiple workflow options.   
```{r}
data(RPPADATA.original)
RPPADATA<-RPPADATA.original
RPPADATA[1:9,1:5]

```
We now have all that is required to run the EUFLOW package. For the purpose of clarity we define the RNASEQDATA set as the EvaluationExperimentSet, which includes 2 workflow options (RnaSeqV1 and RnaSeqv2). We also define the RPPADATA as the ReferenceSet. If it is the choice of the user multiple reference sets can be utilized. 

```{r}

#EvaluationExperimentSet<-RNASEQDATA
EvaluationExperimentSet<-RNASEQV1v2PICCOLO
ReferenceSet<-RPPADATA
```

Now that we have the data for our example, the WorkflowEvaluationData function will modify the separate dataframes into one data structure to prepare to calcuate the model quality and perform the evaluation. The first item in the list is the Reference data and the second item in the list is the evaluation data.


```{r,eval=TRUE}

Workflow.Data<-WorkflowEvaluationData(EvaluationExperimentSet,ReferenceSet)
Workflow.Data[[1]][1:9,1:5]

```
Further data processing to determine the number of workflow options and structure tags will be used to name the output by creating a Merged.options object.

```{r}
Merged.options<-merge_tag_options(Workflow.Data)
Merged.options[1:9,1:5]
dim(Merged.options)
names(Merged.options)
row.names(Merged.options)
```
The Model.quality.object is created to create a map between the Reference ids and the evaluation ids using the Model.quality.list function.

```{r}
Model.quality.object<-Model.quality.list(Merged.options)

```
Next, Model Quality is an object which contains the model quality values for each of the pairs. How the Model quality is determined is specified by the user.
```{r}
Model.Quality<-Workflow.Criterion(Model.quality.object,method="pearson")
head(as.data.frame(Model.Quality))
```
```{r}
Model.Quality<-Workflow.Criterion(Model.quality.object,method="spearman")
head(as.data.frame(Model.Quality))
```

```{r}
Posterior.dataframe<-Workflow.posteriorestimate(Model.quality.object,Model.Quality)

```

```{r}
Evaluation.table<-Workflow.Evaluation.table(Posterior.dataframe)
Evaluation.table
Workflow.Evaluation.table(Posterior.dataframe,deltaPlus = 2)
Workflow.Evaluation.table(Posterior.dataframe,Utp=3)
Workflow.Evaluation.table(Posterior.dataframe,Utp=1)
PlotEvaluationTable(Evaluation.table)

#RNASEQDATAv1$X<-paste(sapply(strsplit(as.character(RNASEQDATAv1$X),"_"), "[", 1),"v3",sep="_")
```

```{r}
data(BRCARPPADATAFINAL)
BRCARPPADATAFINAL[1:9,1:5]
dim(BRCARPPADATAFINAL)
data(BRCASALMONDATAFINAL)
BRCASALMONDATAFINAL[1:9,1:5]
names(BRCASALMONDATAFINAL)
names(BRCARPPADATAFINAL)
names(EvaluationExperimentSet)
names(ReferenceSet)
```

```{r}

#EvaluationExperimentSet<-RNASEQDATA
EvaluationExperimentSet<-BRCASALMONDATAFINAL
ReferenceSet<-BRCARPPADATAFINAL
```

```{r,eval=TRUE}

Workflow.Data<-WorkflowEvaluationData(EvaluationExperimentSet,ReferenceSet)
Workflow.Data[[1]][1:9,1:5]

```

```{r}
Merged.options<-merge_tag_options(Workflow.Data)
Merged.options[1:9,1:5]
dim(Merged.options)
names(Merged.options)
row.names(Merged.options)

Model.quality.object<-Model.quality.list(Merged.options)
Model.Quality<-Workflow.Criterion(Model.quality.object,method="pearson")
head(as.data.frame(Model.Quality))
Posterior.dataframe<-Workflow.posteriorestimate(Model.quality.object,Model.Quality)
Evaluation.table<-Workflow.Evaluation.table(Posterior.dataframe)
Evaluation.table
Workflow.Evaluation.table(Posterior.dataframe,deltaPlus = 2)
Workflow.Evaluation.table(Posterior.dataframe,Utp=3)
Workflow.Evaluation.table(Posterior.dataframe,Utp=1)
PlotEvaluationTable(Evaluation.table)
```

